File: Python/Apply_job.py
import requests,json

baseurl="https://suremdm.42gears.com/api"
Username="Username"
Password="Password"
ApiKey="Your ApiKey"

# Method for getting device info
def GetDeviceID(deviceName):
    '''
    Retreiving information of device
        Endpoint: /device
        Method: POST
        Request Body:
            {  
            "ID": string,
            "SearchValue":string,
            "Limit":integer,
            "SortColumn":string,
            "SortOrder":string,
            "IsSearch":boolean,
            "SearchColumns":string[]
            }
        Authentication:
            Basic authentication
        Headers:
            ApiKey: “Your Api-Key”
    '''
    # Api url
    url = baseurl+"/device"
    # Headers
    headers = {
        # Api-Key header
        'ApiKey': ApiKey,
        # Set Content type
        'Content-Type': "application/json",
        }
    # Basic authentication credentials
    Credentials=(Username,Password)
    # Request body
    PayLoad={
        "ID" : "AllDevices",
        "IsSearch" : bool(1),
        "Limit" : 10,
        "SearchColumns" : ["DeviceName"],
        "SearchValue" : deviceName,
        "SortColumn" : "LastTimeStamp",
        "SortOrder" : "asc"
    }
    # Executing request
    response = requests.post(url,auth=Credentials,json=PayLoad,headers=headers)
    # Extracting required GroupID
    if response.status_code == 200:
        if response.text != '[]':
            data = response.json()
            for group in data['rows']:
                if group['DeviceName'] == deviceName:
                    return group["DeviceID"]
    else:
        return None

# Method for getting ID of the folder
def GetFolderID(folderName):
    '''
    Retrieve ID of the folder
        Endpoint: /jobfolder/all
        Method: GET
        Parameters: NA
        Authentication:
            Basic authentication
        Headers:
            ApiKey: “Your Api-Key”
    '''
    # Api url
    url = baseurl+"/jobfolder/all"
    # Headers
    headers = {
        # Api-Key header
        'ApiKey': ApiKey,
        # Set Content type
        'Content-Type': "application/json",
        }
    # Basic authentication credentials
    Credentials=(Username,Password)
    # Executing request
    response = requests.get(url,auth=Credentials,headers=headers)
    # Extracting required FolderID
    if response.status_code == 200:
        if response.text != '[]':
            data = response.json()
            for folder in data:
                if folder['FolderName'] == folderName:
                    return folder['FolderID']
            return 'null'
    else:
        return "null"

# Method for getting ID of the job
def GetJobID(jobName,folderID):
    '''
    Retrieve ID of the job
        Endpoint: /job
        Method: GET
        Query Parameters: 
            FolderID: ID of the job folder
        Authentication:
            Basic authentication
        Headers:
            ApiKey: “Your Api-Key”
    '''
    # Api url
    url = baseurl+"/job"
    # Headers
    headers = {
        # Api-Key header
        'ApiKey': ApiKey,
        # Set Content type
        'Content-Type': "application/json",
        }
    # Basic authentication credentials
    Credentials=(Username,Password)
    # Query parameters
    params={"FolderID":folderID}
    # Executing request
    response = requests.get(url,auth=Credentials,params=params,headers=headers)
    # Extracting required JobID
    if response.status_code == 200:
        if response.text != '[]':
            data = response.json()
            for job in data:
                if job['JobName'] == jobName:
                    return job['JobID']
    else:
        return None

# Method for applying job
def ApplyJob(jobName,folderName,deviceID):
    '''
    Apply job on device
    Endpoint: /jobassignment
    Method: POST
    Request Body: 
        {
            "DeviceIds": string[],
            "JobId": string
        }
    Authentication:
        Basic authentication
    Headers:
        ApiKey: “Your Api-Key”
    '''
    if folderName!=None:
        FolderID=GetFolderID(folderName)
    else:
        FolderID='null'
    JobID=GetJobID(jobName,FolderID)
    PayLoad={
        "DeviceIds": [deviceID],
        "JobId": JobID
    }
    # Api url
    url = baseurl+"/jobassignment"
    # Headers
    headers = {
        # Api-Key header
        'ApiKey': ApiKey,
        # Set Content type
        'Content-Type': "application/json",
        }
    # Basic authentication credentials
    Credentials=(Username,Password)
    # Executing request
    response = requests.post(url,auth=Credentials,json=PayLoad,headers=headers)
    return response.text

# Main
DeviceID=GetDeviceID("Device_Name")
if DeviceID!=None:
    status=ApplyJob("Job_Name","Folder_Name",DeviceID)
    print(status)
else:
    print('Device not found!')

File: C-Sharp/Apply_Job/Apply_job.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using RestSharp;
using RestSharp.Authenticators;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;

namespace Apply_job
{
    class Apply_job
    {
        public static string baseurl = "https://suremdm.42gears.com/api";
        private static string Username = "Username";
        private static string Password = "Password";
        private static string ApiKey = "Your ApiKey";
        static void Main(string[] args)
        {
            // Applying job on the device requires ID of the Device and ID of the job
            string DeviceID = GetDeviceID("Device_Name");
            if (DeviceID != null)
            {
                ApplyJob("Job_Name", "Jobfolder name", DeviceID);
            }
            else
            {
                Console.WriteLine("Device not found");
            }

            Console.ReadKey();
        }

        // Function to apply job
        private static void ApplyJob(string JobName, string FolderName, string deviceID)
        {
            // Retrieve ID of the jobfolder 
            string FolderID;
            if (FolderName != null)
            {
                FolderID = GetFolderID(FolderName);
            }
            else
            {
                FolderID = null;
            }

            // Pay load for applying job
            var JobApplication = new
            {
                DeviceIds = new List<string>() { deviceID },
                JobId = GetJobID(JobName, FolderID)
            };

            /* Apply job on device
                Endpoint: /jobassignment
                Method: POST
                Request Body: 
                    {
                        "DeviceIds": string[],
                        "JobId": string
                    }
                Authentication:
                    Basic authentication
                Headers:
                    ApiKey: “Your Api-Key”
            */

            string URL = baseurl + "/jobassignment";
            var client = new RestClient(URL);
            //  Basic authentication header
            client.Authenticator = new HttpBasicAuthenticator(Username, Password);
            //  ApiKey Header
            client.AddDefaultHeader("ApiKey", ApiKey);
            //  Set content type
            client.AddDefaultHeader("Content-Type", "application/json");
            // Set request method
            var request = new RestRequest(Method.POST);
            request.AddJsonBody(JobApplication);

            // Execute request
            IRestResponse response = client.Execute(request);
            if (response.StatusCode == HttpStatusCode.OK)
            {
                if (Convert.ToBoolean(response.Content.ToString()))
                {
                    Console.WriteLine("Job applied successfully.");
                }
                else
                {
                    Console.WriteLine("Job application failed!");
                }
            }
            else
            {
                Console.WriteLine(response.Content.ToString());
            }
        }

        // Function to retrieve ID of the job
        private static object GetJobID(string jobName, string folderid)
        {
            /* Retrieve ID of the job
                Endpoint: /job
                Method: GET
                Query Parameters: 
                    FolderID: ID of the job folder
                Authentication:
                    Basic authentication
                Headers:
                    ApiKey: “Your Api-Key”
            */

            string URL = baseurl + "/job";
            var client = new RestClient(URL);
            //  Basic authentication header
            client.Authenticator = new HttpBasicAuthenticator(Username, Password);
            //  ApiKey Header
            client.AddDefaultHeader("ApiKey", ApiKey);
            //  Set content type
            client.AddDefaultHeader("Content-Type", "application/json");
            // Set request method
            var request = new RestRequest(Method.GET);
            request.AddParameter("FolderID", folderid, ParameterType.QueryString);
            //  Execute request
            IRestResponse response = client.Execute(request);
            if (response.StatusCode == HttpStatusCode.OK)
            {
                var OutPut = JsonConvert.DeserializeObject(response.Content);
                var jsonResponse = JsonConvert.DeserializeObject<List<JObject>>(response.Content);
                if ((jsonResponse != null) && (jsonResponse.Any()))
                {
                    foreach (var job in jsonResponse)
                    {
                        if (job.GetValue("JobName").ToString() == jobName)
                        {
                            return job.GetValue("JobID").ToString();
                        }
                    }
                }
            }
            return null;
        }

        // Function to retrieve ID of the jobfolder
        private static string GetFolderID(string folderName)
        {
            /* Retrieve ID of the folder
                Endpoint: /jobfolder/all
                Method: GET
                Parameters: NA
                Authentication:
                    Basic authentication
                Headers:
                    ApiKey: “Your Api-Key”
            */

            string URL = baseurl + "/jobfolder/all";
            var client = new RestClient(URL);
            //  Basic authentication header
            client.Authenticator = new HttpBasicAuthenticator(Username, Password);
            //  ApiKey Header
            client.AddDefaultHeader("ApiKey", ApiKey);
            //  Set content type
            client.AddDefaultHeader("Content-Type", "application/json");
            // Set request method
            var request = new RestRequest(Method.GET);
            //  Execute request
            IRestResponse response = client.Execute(request);
            if (response.StatusCode == HttpStatusCode.OK)
            {
                var OutPut = JsonConvert.DeserializeObject(response.Content);
                List<JObject> jsonResponse = JsonConvert.DeserializeObject<List<JObject>>(response.Content);
                if ((jsonResponse != null) && (jsonResponse.Any()))
                {
                    foreach (var folder in jsonResponse)
                    {
                        if (folder.GetValue("FolderName").ToString() == folderName)
                        {
                            return folder.GetValue("FolderID").ToString();
                        }
                    }
                }
            }
            return "null";
        }

        // Function to retrieve ID of the device
        private static string GetDeviceID(string deviceName)
        {
            /*  Retreiving information of device
                 Endpoint: /device
                 Method: POST
                 Request Body:
                     {  
                        "ID": string,
	                    "SearchValue":string,
	                    "Limit":integer,
	                    "SortColumn":string,
	                    "SortOrder":string,
	                    "IsSearch":boolean,
	                    "SearchColumns":string[]
                    }
                 Authentication:
                     Basic authentication
                 Headers:
                     ApiKey: “Your Api-Key” 
             */

            string URL = baseurl + "/device";
            var client = new RestClient(URL);
            //  Basic authentication header
            client.Authenticator = new HttpBasicAuthenticator(Username, Password);
            //  ApiKey Header
            client.AddDefaultHeader("ApiKey", ApiKey);
            //  Set content type
            client.AddDefaultHeader("Content-Type", "application/json");
            //  Set request method
            var request = new RestRequest(Method.POST);
            //  Request payload
            var RequestPayLoad = new
            {
                ID = "AllDevices",
                IsSearch = true,
                Limit = 10,
                SearchColumns = new string[] { "DeviceName" },
                SearchValue = deviceName,
                SortColumn = "LastTimeStamp",
                SortOrder = "asc"
            };
            request.AddJsonBody(RequestPayLoad);

            //  Execute request
            IRestResponse response = client.Execute(request);

            if (response.StatusCode == HttpStatusCode.OK)
            {
                var OutPut = JsonConvert.DeserializeObject<JObject>(response.Content);
                foreach (var device in OutPut["rows"])
                {
                    if ((string)device["DeviceName"] == deviceName)
                    {
                        return device["DeviceID"].ToString();
                    }
                }
            }
            return null;
        }
    }
}
